File: Programmers/python/성격 유형 검사하기.py
# 2022 KAKAO TECH INTERNSHIP - 성격 유형 검사하기

def find(personal, type1, type2):
    weight1 = personal[type1]
    weight2 = personal[type2]
    
    if weight1 == weight2:
        return type1 if type1 < type2 else type2
    
    return type1 if weight1 > weight2 else type2
    

def solution(survey, choices):
    answer = ''
    personal = {'R':0, 'T':0, 'C':0, 'F':0, 'J':0, 'M':0, 'A':0, 'N':0}
    weight = [-3, -2, -1, 0, 1, 2, 3]
    length = len(survey)
    
    for i in range(0, length):
        negative = survey[i][0]
        positive = survey[i][1]
        choice = choices[i] - 1
        
        if weight[choice] > 0:
            personal[positive] = personal[positive] + weight[choice]
        else: 
            personal[negative] = personal[negative] + (weight[choice] * -1)
    
    answer += find(personal, 'R', 'T')
    answer += find(personal, 'C', 'F')
    answer += find(personal, 'J', 'M')
    answer += find(personal, 'A', 'N')
    
    return answer

File: Programmers/java/성격 유형 검사하기.java
import java.util.*;

class Solution {
    Map<String, Integer> weight;
    
    public String solution(String[] survey, int[] choices) {
        String answer = "";
        weight = init();
        
        for(int i = 0; i < survey.length; i++) {
            String negative = String.valueOf(survey[i].charAt(0));
            String positive = String.valueOf(survey[i].charAt(1));
            int choice = choices[i] - 4;
            
            if (choice > 0) {
                weight.compute(positive, (key, value) -> value + Math.abs(choice));
            }
            else {
                weight.compute(negative, (key, value) -> value + Math.abs(choice));
            }
            
        }
        answer += calculate("R", "T");
        answer += calculate("C", "F");
        answer += calculate("J", "M");
        answer += calculate("A", "N");

        return answer;
    }
    
    public String calculate(String type1, String type2) {
        int weight1 = weight.get(type1);
        int weight2 = weight.get(type2);
        
        if(weight1 == weight2) {
            if(type1.compareTo(type2) > 0) {
                return type2;
            }
            return type1;
        }
        return weight1 > weight2 ? type1 : type2;
    }
    
    public Map<String, Integer> init() {
        Map<String, Integer> weight = new HashMap<>();
        weight.put("R", 0);
        weight.put("T", 0);
        weight.put("C", 0);
        weight.put("F", 0);
        weight.put("J", 0);
        weight.put("M", 0);
        weight.put("A", 0);
        weight.put("N", 0);
        return weight;
    }
}
